File: gtci/04-merge-intervals/python3/5-maximum-cpu-load.py
import heapq


class Job:
    def __init__(self, start, end, cpu_load):
        self.start = start
        self.end = end
        self.cpuLoad = cpu_load


class Solution:
    def findMaxCPULoad(self, jobs):
        if len(jobs) < 2:
            return 0 if len(jobs) == 0 else jobs[0].cpuLoad

        jobs.sort(key=lambda x: x.start)

        max_cpu_load = 0

        current_cpu_load = 0
        min_heap = []

        for job in jobs:
            while min_heap and min_heap[0][1] <= job.start:
                popped_job = heapq.heappop(min_heap)
                current_cpu_load -= popped_job[2]

            heapq.heappush(min_heap, (job.start, job.end, job.cpuLoad))
            current_cpu_load += job.cpuLoad

            max_cpu_load = max(max_cpu_load, current_cpu_load)

        return max_cpu_load


if __name__ == "__main__":
    sol = Solution()

    assert sol.findMaxCPULoad([Job(1, 4, 3), Job(2, 5, 4), Job(7, 9, 6)]) == 7
    assert sol.findMaxCPULoad([Job(6, 7, 2), Job(2, 4, 11), Job(8, 12, 15)]) == 15
    assert sol.findMaxCPULoad([Job(1, 4, 2), Job(2, 4, 1), Job(3, 6, 5)]) == 8

    print("All test cases passed.")

File: gtci/04-merge-intervals/cpp/5-maximum-cpu-load.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Job {
   public:
    int start = 0;
    int end = 0;
    int cpuLoad = 0;

    Job(int start, int end, int cpuLoad) {
        this->start = start;
        this->end = end;
        this->cpuLoad = cpuLoad;
    }
};

class Solution {
   public:
    int findMaxCPULoad(vector<Job>&& jobs) {
        if (jobs.size() == 0) return 0;

        if (jobs.size() == 1) return jobs[0].cpuLoad;

        std::sort(jobs.begin(), jobs.end(),
                  [](Job a, Job b) { return a.start < b.start; });

        auto jobCmp = [](const Job& a, const Job& b) { return a.end > b.end; };
        std::priority_queue<Job, std::vector<Job>, decltype(jobCmp)> minHeap(
            jobCmp);

        int maxCpuLoad = 0;
        int currentCpuLoad = 0;

        for (Job& job : jobs) {
            while (!minHeap.empty() && minHeap.top().end <= job.start) {
                const Job& poppedJob = minHeap.top();
                minHeap.pop();
                currentCpuLoad -= poppedJob.cpuLoad;
            }

            minHeap.push(job);
            currentCpuLoad += job.cpuLoad;

            maxCpuLoad = std::max(maxCpuLoad, currentCpuLoad);
        }

        return maxCpuLoad;
    }
};

int main() {
    Solution sol;

    assert(sol.findMaxCPULoad({Job(1, 4, 3), Job(2, 5, 4), Job(7, 9, 6)}) == 7);
    assert(sol.findMaxCPULoad({Job(6, 7, 2), Job(2, 4, 11), Job(8, 12, 15)}) ==
           15);
    assert(sol.findMaxCPULoad({Job(1, 4, 2), Job(2, 4, 1), Job(3, 6, 5)}) == 8);

    cout << "All test cases passed." << endl;

    return 0;
}
